### PR TITLE
feat(pipelines): add ComponentId struct

### PR DESCRIPTION
**Pull request by jdrouet**
Signed-off-by: Jérémie Drouet <jeremie.drouet@datadoghq.com>

### CODE REVIEW

File: src/api/schema/components/mod.rs [hunk 1]
```diff
@@ -203,9 +204,11 @@ impl ComponentsQuery {
         .await
     }
 
-    /// Gets a configured component by name
-    async fn component_by_name(&self, name: String) -> Option<Component> {
-        component_by_name(&name)
+    /// Gets a configured component by id
+    async fn component_by_id(&self, name: String) -> Option<Component> {
```
Comment (jdrouet) [inline src/api/schema/components/mod.rs:208]: for backward compatibility, we might need to consider having 2 functions:
- `component_by_name` which would _only_ give global components
- `component_by_id` which would allow to select by `ComponentId`.

File: src/config/mod.rs [hunk 2]
```diff
@@ -72,17 +72,67 @@ impl<'a> From<&'a ConfigPath> for &'a PathBuf {
...
+        write!(f, "{:?}", self.name)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct WithInputs<Input, Inner> {
+    pub inputs: Vec<Input>,
+    #[serde(flatten)]
+    pub inner: Inner,
+}
```
Comment (lukesteensen) [inline src/config/mod.rs:99]: I'm a bit skeptical of this helper type since it seems to add a decent amount of indirection around something relatively simple. What if instead we allowed a `ComponentId` to be deserialized from a raw string? That might be just as convenient and quite a bit simpler.

File: src/config/mod.rs [hunk 3]
```diff
@@ -466,15 +515,47 @@ impl Display for Resource {
 
 #[derive(Deserialize, Serialize, Debug)]
 #[serde(deny_unknown_fields)]
-pub struct TestDefinition {
+pub struct TestDefinition<Id> {
```
Comment (lukesteensen) [inline src/config/mod.rs:518]: It feels strange to make this generic. It seems like it's mostly to allow for deserialization with `String` keys. As mentioned above, maybe all this can go away if we deserialize strings directly into component IDs?

File: src/config/unit_test.rs [hunk 4]
```diff
@@ -189,7 +201,7 @@ impl UnitTest {
             if let Some((inputs, outputs)) = results.get(&check.extract_from) {
                 if check.conditions.is_empty() {
                     inspections.push(format!(
-                        "check transform '{}' payloads (events encoded as JSON):\n{}\n{}",
+                        "check transform {:?} payloads (events encoded as JSON):\n{}\n{}",
```
Comment (lukesteensen) [inline src/config/unit_test.rs:204]: We're using `Debug` quite a bit for error messages and I'm not sure it's the best choice. How about we stick with `Display` and add an explicit implementation for `ComponentId` so that it's a bit more clear this is user-facing?

File: src/topology/running.rs [hunk 5]
```diff
@@ -321,18 +329,24 @@ impl RunningTopology {
...
+                .sources
+                .get(id)
+                .map(|source| (id, source.inner.resources()))
+        });
+        let add_sink = diff.sinks.changed_and_added().filter_map(|id| {
+            new_config
+                .sinks
+                .get(id)
+                .map(|sink| (id, sink.inner.resources(&id.name)))
+        });
```
Comment (lukesteensen) [inline src/topology/running.rs:349]: This looks like it could be a change in behavior (i.e. `map` to `filter_map` and indexing to `get`). If so, is that intentional?

File: src/topology/running.rs [hunk 6]
```diff
@@ -347,15 +361,18 @@ impl RunningTopology {
...
         // We can reuse buffers whose configuration wasn't changed.
         let reuse_buffers = diff
             .sinks
             .to_change
             .iter()
-            .filter(|&name| self.config.sinks[name].buffer == new_config.sinks[name].buffer)
+            .filter(|&id| {
+                self.config.sinks.get(id).map(|item| &item.inner.buffer)
+                    == new_config.sinks.get(id).map(|item| &item.inner.buffer)
+            })
```
Comment (lukesteensen) [inline src/topology/running.rs:375]: Here as well, I think. Previously we'd blow up if something is missing that we expect to be present, and now it'll pass silently. Likely not a huge deal, just curious what the motivation is behind changing the behavior.

File: src/topology/running.rs [hunk 7]
```diff
@@ -399,20 +416,20 @@ impl RunningTopology {
...
         // Cleanup changed and collect buffers to be reused
         let mut buffers = HashMap::new();
-        for name in &diff.sinks.to_change {
-            if wait_for_sinks.contains(name) {
-                let previous = self.tasks.remove(name).unwrap();
-                debug!(message = "Waiting for sink to shutdown.", %name);
+        for id in &diff.sinks.to_change {
+            if wait_for_sinks.contains(id) {
+                let previous = self.tasks.remove(id).unwrap();
+                debug!(message = "Waiting for sink to shutdown.", name = ?id.name);
```
Comment (lukesteensen) [inline src/topology/running.rs:422]: A `Display` implementation for `ComponentId` would also mean none of these have to change.

File: src/topology/running.rs [hunk 8]
```diff
@@ -423,81 +440,87 @@ impl RunningTopology {
...
+            // TODO handle sending ids
             self.watch
                 .0
-                .send(self.outputs.clone())
+                .send(
+                    self.outputs
+                        .iter()
+                        .map(|item| (item.0.name.clone(), item.1.clone()))
+                        .collect::<HashMap<_, _>>(),
+                )
```
Comment (lukesteensen) [inline src/topology/running.rs:481]: Is this TODO something that needs handled before merging?

**Review by lukesteensen — COMMENTED** [review]
Overall this is looking good! Just a few things to consider:
1. A `Deserialize` impl for `ComponentId` from a string seems like it could simplify this a good bit.
2. A `Display` impl for `ComponentId` seems like it'd reduce the diff significantly and make it clearer what representation is user-facing
3. A couple of small-ish behavior changes that we might want to either skip or document the reasoning

File: src/config/mod.rs (hunk 2 above)
Comment (jdrouet) [reply src/config/mod.rs:99]: yeah, but when de-serializing a pipeline, the `ComponentId` wont be able to know that it's in the `pipeline_name.toml` and therefore set the good `scope` value. But I guess this can be done in a second step.

File: src/topology/running.rs (hunk 5 above)
Comment (jdrouet) [reply src/topology/running.rs:349]: The `indexing` to `get` is due to the fact that we cannot use a struct for indexing (or I don't know about it and I'll look for it).
The `map` to `filter_map` is due to the `get` returning an `Option` but the behavior should be the same in the end.

File: src/topology/running.rs (hunk 6 above)
Comment (jdrouet) [reply src/topology/running.rs:375]: same problem, indexing with a struct 🤔

File: src/api/schema/components/mod.rs (hunk 1 above)
Comment (leebenson) [reply src/api/schema/components/mod.rs]: Just noting the likely merge conflicts with https://github.com/timberio/vector/pull/8417 (merged 13 hours ago) which refactored the component name field to `component_by_component_id`.

Any API changes also need reflecting in the API clients, including `tap` and `top`. #8417 shows a good example of where code was changed to accommodate the client features.

To ensure the API and its client remain in sync, run `make build-graphql-schema` after making API changes and then attempt to `cargo build` again. The compiler should alert you to anything that's fallen out of sync.

File: src/config/mod.rs (hunk 2 above)
Comment (jdrouet) [reply src/config/mod.rs]: I applied your solution, looks great 👍

File: src/topology/running.rs (hunk 5 above)
Comment (jdrouet) [reply src/topology/running.rs]: Ok, works with indexing with a struct, don't know why I changed that...

File: src/config/id.rs [hunk 9]
```diff
@@ -0,0 +1,108 @@
+use serde::de::{self, Visitor};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::cmp::{Ord, Ordering, PartialOrd};
+use std::fmt;
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ComponentId {
+    pub name: String,
```
Comment (jszwedko) [inline src/config/id.rs:8]: I know this is a bit repetitive with the struct name, but I might rename this field `id` to match how we switched to referring to component IDs in https://github.com/timberio/vector/pull/8417

**Comment by leebenson** [comment]
@jdrouet, running `cargo check` locally yields:

```
error[E0559]: variant `vector_core::buffers::Variant::Disk` has no field named `id`
  --> src/buffers.rs:68:17
   |
68 |                 id: sink_id.to_string(),
   |                 ^^ `vector_core::buffers::Variant::Disk` does not have this field
   |
   = note: available fields are: `max_size`, `when_full`, `data_dir`, `name`
```

Are you seeing this?

Doesn't seem to have been picked up by CI and oddly, `cargo build` locally actually succeeds ❓

**Comment by jdrouet** [comment]
@leebenson I don't have this locally. I'll try a `cargo clean` first and come back 😉

**Review by spencergilbert — APPROVED** [review]

File: src/api/schema/components/mod.rs [hunk 10]
```diff
@@ -39,11 +40,11 @@ pub enum ComponentKind {
...
 impl Component {
-    fn get_component_id(&self) -> &str {
+    fn get_component_id(&self) -> String {
         match self {
-            Component::Source(c) => c.0.component_id.as_str(),
-            Component::Transform(c) => c.0.component_id.as_str(),
-            Component::Sink(c) => c.0.component_id.as_str(),
+            Component::Source(c) => c.0.component_id.to_string(),
+            Component::Transform(c) => c.0.component_id.to_string(),
+            Component::Sink(c) => c.0.component_id.to_string(),
```
Comment (leebenson) [inline src/api/schema/components/mod.rs:47]: Nit: Since `ComponentId.id` is public, we could avoid the allocation with:

```rust
fn get_component_id(&self) -> &str {
    match self {
        Component::Source(c) => c.0.component_id.id.as_str(),
        Component::Transform(c) => c.0.component_id.id.as_str(),
        Component::Sink(c) => c.0.component_id.id.as_str(),
    }
}
```

Alternatively, if it's worth hiding the inner `id` for whatever reason, it could follow the `String::as_str()` implementation with:

```rust
impl ComponentId {
    pub fn as_str(&self) -> &str {
        &self.id
    }
}
```

File: src/api/schema/components/sink.rs [hunk 11]
```diff
@@ -5,25 +5,27 @@ use crate::{
...
+    pub inputs: Vec<ComponentId>,
 }
 
 #[derive(Debug, Clone)]
 pub struct Sink(pub Data);
 
 impl Sink {
-    pub fn get_component_id(&self) -> &str {
-        self.0.component_id.as_str()
+    pub fn get_component_id(&self) -> &ComponentId {
```
Comment (leebenson) [inline src/api/schema/components/sink.rs:25]: It's not obvious to me why `Component.get_component_id` returns a `String` (or a `&str` per my previous suggestion), yet a sink returns a `&ComponentId`.

Could these two be standardised?

File: src/api/schema/components/sink.rs [hunk 12]
```diff
@@ -39,9 +41,9 @@ pub struct SinksFilter {
 impl filter::CustomFilter<Sink> for SinksFilter {
     fn matches(&self, sink: &Sink) -> bool {
         filter_check!(
-            self.component_id
-                .as_ref()
-                .map(|f| f.iter().all(|f| f.filter_value(sink.get_component_id()))),
+            self.component_id.as_ref().map(|f| f
+                .iter()
+                .all(|f| f.filter_value(&sink.get_component_id().to_string()))),
```
Comment (leebenson) [inline src/api/schema/components/sink.rs:46]: This is another area where I think it'd make sense to deal in `&str` rather than having to allocate a new string to filter.

File: src/api/schema/components/sink.rs [hunk 13]
```diff
@@ -63,9 +65,10 @@ pub enum SinksSortFieldName {
 impl sort::SortableByField<SinksSortFieldName> for Sink {
     fn sort(&self, rhs: &Self, field: &SinksSortFieldName) -> cmp::Ordering {
         match field {
-            SinksSortFieldName::ComponentId => {
-                Ord::cmp(self.get_component_id(), rhs.get_component_id())
-            }
+            SinksSortFieldName::ComponentId => Ord::cmp(
+                &self.get_component_id().to_string(),
```
Comment (leebenson) [inline src/api/schema/components/sink.rs:69]: Same as above.

File: src/api/schema/components/sink.rs [hunk 14]
```diff
@@ -76,8 +79,8 @@ impl sort::SortableByField<SinksSortFieldName> for Sink {
 #[Object]
 impl Sink {
     /// Sink component_id
-    pub async fn component_id(&self) -> &str {
-        self.get_component_id()
+    pub async fn component_id(&self) -> String {
+        self.get_component_id().to_string()
```
Comment (leebenson) [inline src/api/schema/components/sink.rs:83]: Same as above.

File: src/api/schema/components/source.rs [hunk 15]
```diff
@@ -77,8 +78,8 @@ impl sort::SortableByField<SourcesSortFieldName> for Source {
 #[Object]
 impl Source {
     /// Source component_id
-    pub async fn component_id(&self) -> &str {
-        self.get_component_id()
+    pub async fn component_id(&self) -> String {
+        self.get_component_id().to_string()
```
Comment (leebenson) [inline src/api/schema/components/source.rs:82]: As with `Sink`, I'd retain the previous pattern of returning a `&str`.

File: src/api/schema/components/source.rs [hunk 16]
```diff
@@ -125,9 +126,9 @@ pub struct SourcesFilter {
 impl filter::CustomFilter<Source> for SourcesFilter {
     fn matches(&self, source: &Source) -> bool {
         filter_check!(
-            self.component_id
-                .as_ref()
-                .map(|f| f.iter().all(|f| f.filter_value(source.get_component_id()))),
+            self.component_id.as_ref().map(|f| f
+                .iter()
+                .all(|f| f.filter_value(&source.get_component_id().to_string()))),
```
Comment (leebenson) [inline src/api/schema/components/source.rs:131]: Same as above.

File: src/api/schema/components/transform.rs [hunk 17]
```diff
@@ -5,24 +5,25 @@ use crate::{
...
+    pub inputs: Vec<ComponentId>,
 }
 
 #[derive(Debug, Clone)]
 pub struct Transform(pub Data);
 
 impl Transform {
-    pub fn get_component_id(&self) -> &str {
-        self.0.component_id.as_str()
+    pub fn get_component_id(&self) -> String {
```
Comment (leebenson) [inline src/api/schema/components/transform.rs:25]: Another opportunity for `&str`.

File: src/api/schema/components/transform.rs [hunk 18]
```diff
@@ -51,7 +52,7 @@ impl sort::SortableByField<TransformsSortFieldName> for Transform {
 #[Object]
 impl Transform {
     /// Transform component_id
-    pub async fn component_id(&self) -> &str {
+    pub async fn component_id(&self) -> String {
```
Comment (leebenson) [inline src/api/schema/components/transform.rs:55]: As above.

File: src/api/schema/metrics/filter.rs [hunk 19]
```diff
@@ -144,9 +145,12 @@ pub fn get_all_metrics(interval: i32) -> impl Stream<Item = Vec<Metric>> {
 }
 
 /// Return Vec<Metric> based on a component id tag.
-pub fn by_component_id(component_id: &str) -> Vec<Metric> {
+pub fn by_component_id(component_id: &ComponentId) -> Vec<Metric> {
     capture_metrics(&GLOBAL_CONTROLLER)
-        .filter_map(|m| m.tag_matches("component_id", component_id).then(|| m))
+        .filter_map(|m| {
+            m.tag_matches("component_id", &component_id.to_string())
```
Comment (leebenson) [inline src/api/schema/metrics/filter.rs:151]: As above.

File: src/config/compiler.rs [hunk 20]
```diff
@@ -32,7 +32,7 @@ pub fn compile(mut builder: ConfigBuilder) -> Result<(Config, Vec<String>), Vec<
                 sources: builder.sources,
                 sinks: builder.sinks,
                 transforms: builder.transforms,
-                tests: builder.tests,
+                tests: builder.tests.into_iter().map(Into::into).collect(),
```
Comment (leebenson) [inline src/config/compiler.rs:35]: Why was this change needed?

File: src/topology/running.rs [hunk 21]
```diff
@@ -453,9 +461,15 @@ impl RunningTopology {
 
         // Broadcast changes to subscribers.
         if !self.watch.0.is_closed() {
+            // TODO handle sending ids
```
Comment (lukesteensen) [inline src/topology/running.rs:464]: If we're not going to do this as part of the PR then it's probably a good idea to open an issue so we don't forget.

File: src/validate.rs [hunk 22]
```diff
@@ -222,23 +222,22 @@ async fn validate_healthchecks(
...
                 if config
                     .sinks
-                    .get(&name)
+                    .get(&id)
                     .expect("Sink not present")
                     .healthcheck()
                     .enabled
                 {
-                    fmt.success(format!("Health check `{}`", name.as_str()));
+                    fmt.success(format!("Health check {:?}", id));
```
Comment (lukesteensen) [inline src/validate.rs:230]:
```suggestion
                    fmt.success(format!("Health check `{}`", id));
```

File: src/validate.rs [hunk 23]
```diff
@@ -222,23 +222,22 @@ async fn validate_healthchecks(
...
                 } else {
-                    fmt.warning(format!("Health check disabled for `{}`", name));
+                    fmt.warning(format!("Health check disabled for {:?}", id));
```
Comment (lukesteensen) [inline src/validate.rs:232]:
```suggestion
                    fmt.warning(format!("Health check disabled for `{}`", id));
```

File: src/validate.rs [hunk 24]
```diff
@@ -222,23 +222,22 @@ async fn validate_healthchecks(
...
-            Err(error) if error.is_cancelled() => failed(format!(
-                "Health check for `{}` was cancelled",
-                name.as_str()
-            )),
-            Err(_) => failed(format!("Health check for `{}` panicked", name.as_str())),
+            Ok(Err(())) => failed(format!("Health check for {:?} failed", id)),
+            Err(error) if error.is_cancelled() => {
+                failed(format!("Health check for {:?} was cancelled", id))
+            }
+            Err(_) => failed(format!("Health check for {:?} panicked", id)),
```
Comment (lukesteensen) [inline src/validate.rs:240]:
```suggestion
            Ok(Err(())) => failed(format!("Health check for `{}` failed", id)),
            Err(error) if error.is_cancelled() => {
                failed(format!("Health check for `{}` was cancelled", id))
            }
            Err(_) => failed(format!("Health check for `{}` panicked", id)),
```

**Review by lukesteensen — APPROVED** [review]
A few smaller suggestions, but overall this looks good! Thanks @jdrouet

File: src/config/id.rs [hunk 25]
```diff
@@ -0,0 +1,106 @@
+use serde::de::{self, Visitor};
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::cmp::{Ord, Ordering, PartialOrd};
+use std::fmt;
```
Comment (leebenson) [inline src/config/id.rs:4]: Nit:

```suggestion
use serde::{
    de::{self, Visitor},
    Deserialize, Deserializer, Serialize, Serializer,
};
use std::{
    cmp::{Ord, Ordering, PartialOrd},
    fmt,
};
```

File: src/config/id.rs [hunk 26]
```diff
@@ -0,0 +1,106 @@
...
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ComponentId {
+    pub id: String,
```
Comment (leebenson) [inline src/config/id.rs:8]: Does this need to be public?

File: src/api/schema/components/mod.rs [hunk 27]
```diff
@@ -68,7 +69,7 @@ impl filter::CustomFilter<Component> for ComponentsFilter {
         filter_check!(
             self.component_id.as_ref().map(|f| f
                 .iter()
-                .all(|f| f.filter_value(component.get_component_id()))),
+                .all(|f| f.filter_value(&component.get_component_id()))),
```
Comment (leebenson) [inline src/api/schema/components/mod.rs:72]: The borrow here can be dropped if the suggestion to return `&str` above is used.

File: src/sources/file.rs [hunk 28]
```diff
@@ -210,7 +210,8 @@ impl SourceConfig for FileConfig {
         // other
         let data_dir = cx
             .globals
-            .resolve_and_make_data_subdir(self.data_dir.as_ref(), &cx.id)?;
+            // source are only global, name can be used for subdir
+            .resolve_and_make_data_subdir(self.data_dir.as_ref(), &cx.id.to_string())?;
```
Comment (leebenson) [inline src/sources/file.rs:214]: `resolve_and_make_data_subdir` takes a &str; this could be satisfied with `&cx.id.id` in the current public context, or `cx.id.as_str()` per previous suggestions, and avoid the allocation.

File: src/sources/journald.rs [hunk 29]
```diff
@@ -98,7 +98,8 @@ impl SourceConfig for JournaldConfig {
 
         let data_dir = cx
             .globals
-            .resolve_and_make_data_subdir(self.data_dir.as_ref(), &cx.id)?;
+            // source are only global, name can be used for subdir
+            .resolve_and_make_data_subdir(self.data_dir.as_ref(), &cx.id.to_string())?;
```
Comment (leebenson) [inline src/sources/journald.rs:102]: As above.

File: src/sources/kubernetes_logs/mod.rs [hunk 30]
```diff
@@ -217,7 +217,8 @@ impl Source {
         };
         let client = k8s::client::Client::new(k8s_config, proxy)?;
 
-        let data_dir = globals.resolve_and_make_data_subdir(config.data_dir.as_ref(), name)?;
+        let data_dir =
+            globals.resolve_and_make_data_subdir(config.data_dir.as_ref(), &id.to_string())?;
```
Comment (leebenson) [inline src/sources/kubernetes_logs/mod.rs:221]: As above.

**Review by leebenson — CHANGES_REQUESTED** [review]
Looking good @jdrouet.

Just a few comments on avoiding some string allocations.

I think that's worth doing, since these are potentially hot paths in high-volume API metrics when the refresh internal is fairly low.

I tested `top` and `tap` and both seem to work well with the new component ID 👍🏻

File: src/api/schema/components/mod.rs (hunk 10 above)
Comment (jdrouet) [reply src/api/schema/components/mod.rs:47]: This conversion to `&str` works well for now, be if we look ahead and have the following

```rust
let component_id = ComponentId {
    id: "foo".to_string(),
    scope: Scope::Pipeline("bar".to_string()),
};
```
The `component_id.id.as_str()` will probably not be the same than `component_id.to_string()`.
We might have to use the `bar#foo` representation.

File: src/api/schema/components/sink.rs (hunk 11 above)
Comment (jdrouet) [reply src/api/schema/components/sink.rs:25]: Agreed, I think I forgot the other one. 👍

File: src/api/schema/components/mod.rs (hunk 10 above)
Comment (leebenson) [reply src/api/schema/components/mod.rs:47]: I think this could be created _once_ at the time of `ComponentId` construction, and then `.as_str()` would reference that 'cached' version. That'd make allocation a one-time cost.

In either case, it'd remain an implementation detail -- the caller doesn't need to consider what `.as_str()` returns.

What do you think?

Comment (jdrouet) [reply src/api/schema/components/mod.rs:47]: That seems fair to me to have both `inner: String` and `id: String`. I don't really know if having one more `String` in memory or doing one more allocation is better though.

Comment (leebenson) [reply src/api/schema/components/mod.rs:47]: I'd personally opt for `.as_str()`, and create a `ComponentId::new()` method to hide the implementation details of a later `inner` cache.

That would avoid unnecessary allocations for now, when just an `id` is present.

And in future, that would reduce the cost of formatting to a single allocation (at least where taking ownership of it is unnecessary.)

It seems like in most code paths, at least one call to `.to_string()` is present, so having this cost at the time of creation vs. later doesn't seem any more expensive. And it's less expensive in any code path where we're taking the reference to the string twice -- which seems often enough for this to be worth it.

It's minor overall, but I think it's worth doing for high-volume API queries where those string allocations can add up (filtering, etc.)

File: src/config/id.rs [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+    fn from(id: String) -> Self {
+        Self { id }
+    }
+}
+
+impl From<&str> for ComponentId {
+    fn from(id: &str) -> Self {
+        Self { id: id.to_string() }
+    }
+}
```
[the remark this answers is not in the document]
Comment (jdrouet) [reply src/config/id.rs:27]: According to rustc, `impl<T: ToString> From<T> for ComponentId` is conflicting with `impl From<T> to T`, so apparently it's not possible because `ComponentId` implements `ToString`.

File: src/topology/running.rs (hunk 21 above)
Comment (jdrouet) [reply src/topology/running.rs]: I implemented what was needed 😉

**Review by leebenson — APPROVED** [review]
🎉

File: src/config/id.rs [hunk 32]
```diff
@@ -0,0 +1,122 @@
...
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+use std::{
+    cmp::{Ord, Ordering, PartialOrd},
+    fmt,
+};
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash)]
+pub struct ComponentId {
+    value: String,
```
Comment (jszwedko) [inline src/config/id.rs:12]: @jdrouet Sorry for this late comment, but I'm wondering what `value` is?

Comment (leebenson) [reply src/config/id.rs:12]: I suspect I inadvertently prompted this by suggesting an `.as_str()` implementation.

I don't think this is currently needed. The idea being, we'd introduce an 'inner' value if/when the `id` field is augmented with a secondary field for scope. At that time, the inner value would represent a pre-allocated name for the full id + scope, and we'd avoid multiple to_string() calls.

In the interim, `.as_str()` should just pull from id.

I'll fire off a quick follow-up PR for this small adjustment.

Comment (jdrouet) [reply src/config/id.rs:12]: My bad, I prepared this for the next PR adding the scope.
